cmd/libsnap-confine-private/cgroup-support.c

```diff
@@ -97,3 +100,57 @@ bool sc_cgroup_is_v2() {
     }
     return false;
 }
+
+void sc_cgroup_mount_snapd_hierarchy(void) {
+    struct stat stat_buf;
+    struct statfs statfs_buf;
+
+    if (lstat(snapd_cgroup_dir, &stat_buf) < 0) {
+        if (errno != ENOENT) {
+            die("cannot stat %s", snapd_cgroup_dir);
+        }
+        if (statfs(cgroup_dir, &statfs_buf) < 0) {
+            die("cannot fstfs %s", cgroup_dir);
+        }
+        if (statfs_buf.f_flags & MS_RDONLY) {
+            // Remount /sys/fs/cgroup read-write if it was read-only. It's
+            // normally switched read-only after systemd is done initializing.
+            if (mount("none", cgroup_dir, "", MS_REMOUNT, NULL) < 0) {
+                die("cannot remount %s as writable", cgroup_dir);
+            }
+        }
+        if (mkdir(snapd_cgroup_dir, 0555) < 0) {
+            die("cannot mkdir %s", snapd_cgroup_dir);
+        }
+        if (chown(snapd_cgroup_dir, 0, 0) < 0) {
+            die("cannot chown %s to root.root", snapd_cgroup_dir);
+        }
+        if (statfs_buf.f_flags & MS_RDONLY) {
+            if (mount("none", cgroup_dir, "", MS_REMOUNT | MS_RDONLY, NULL) < 0) {
+                die("cannot remount %s as read-only", cgroup_dir);
+            }
+        }
+    } else {
+        if ((stat_buf.st_mode & S_IFMT) != S_IFDIR) {
+            die("cannot proceed: %s must be a directory", snapd_cgroup_dir);
+        }
+    }
+
+    if (statfs(snapd_cgroup_dir, &statfs_buf) < 0) {
+        die("cannot fstfs %s", snapd_cgroup_dir);
+    }
+    if (statfs_buf.f_type != CGROUP_SUPER_MAGIC) {
+        // The "none" indicates that no controllers should be enabled there.
+        if (mount("cgroup", snapd_cgroup_dir, "cgroup", MS_NOSUID | MS_NODEV | MS_NOEXEC | MS_RELATIME,
+                  "none,name=snapd") < 0) {
+            if (errno == ENOENT) {
+                die("cannot mount snapd cgroup v1 hierarchy, disabled in the kernel?");
```

Comment:
Shouldn't we try to degrade gracefully here?

Reply:
Not sure yet, this is early code still.

Reply:
I ended up not making this graceful for the following reason. The original pid tracking was not graceful either. Before we need to consider this thing failing and other parts of snapd adjusting I would rather not to, because it adds additional complexity as to what the runtime state is. We can revisit this ahead of the public release.